tools/gradle-plugin/src/main/kotlin/io/github/composegears/valkyrie/gradle/Utils.kt

```diff
@@ -0,0 +1,67 @@
+package io.github.composegears.valkyrie.gradle
+
+import com.android.build.api.dsl.CommonExtension
+import org.gradle.api.GradleException
+import org.gradle.api.Project
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.file.Directory
+import org.gradle.api.file.FileCollection
+import org.gradle.api.provider.Provider
+import org.gradle.util.GradleVersion
+import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
+
+private val NO_PACKAGE_NAME_ERROR = """
+    Couldn't automatically estimate package name - make sure to set this property in your gradle script like:
+
+    valkyrie {
+        packageName = "my.output.package.name"
+    }
+""".trimIndent()
+
+internal fun ConfigurableFileCollection.conventionCompat(
+    paths: Iterable<*>,
+): ConfigurableFileCollection = if (GradleVersion.current() >= GradleVersion.version("8.8")) {
+    @Suppress("UnstableApiUsage")
+    convention(paths)
+} else {
+    setFrom(paths)
+    this
+}
+
+private val ANDROID_PLUGIN_IDS = listOf(
+    "com.android.application",
+    "com.android.library",
+    "com.android.test",
+    "com.android.dynamic-feature",
+)
+
+internal fun Project.packageNameOrThrow(): Provider<String?> = provider {
+    if (ANDROID_PLUGIN_IDS.any(pluginManager::hasPlugin)) {
+        extensions
+            .findByType(CommonExtension::class.java)
+            ?.namespace
+            ?: throw GradleException(NO_PACKAGE_NAME_ERROR)
+    } else {
+        throw GradleException(NO_PACKAGE_NAME_ERROR)
+    }
+}
+
+internal fun Project.defaultOutputDir(sourceSet: KotlinSourceSet) = project
+    .layout
+    .buildDirectory
+    .dir("generated/sources/valkyrie/${sourceSet.name}")
+
+internal fun KotlinSourceSet.root(): Directory = project
+    .layout
+    .projectDirectory
+    .dir("src/$name")
```

Comment:
I guess you need

```kt
org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet#getKotlin
org.gradle.api.file.SourceDirectorySet#getSrcDirs

kotlin.srcDirs
```

or related things.

Reply:
I can call `sourceSet.kotlin.srcDirs` which gives the kotlin and java folders, but there's no benefit to doing that instead of what I've got already

Reply:
`"src/$name"` is hardcoding, this dir could be changed by users.

Reply:
Done I think? Can the user change the "src" dir's path though?